internal/evmreader/evmreader_test.go

```diff
@@ -0,0 +1,709 @@
+
+//go:embed testdata/input_added_event_2.json
+var inputAddedEvent2JsonBytes []byte
+
+//go:embed testdata/input_added_event_3.json
+var inputAddedEvent3JsonBytes []byte
+
+//go:embed testdata/header_0.json
+var header0JsonBytes []byte
+
+//go:embed testdata/header_1.json
+var header1JsonBytes []byte
+
+//go:embed testdata/header_2.json
+var header2JsonBytes []byte
+
+var (
+	header0 = types.Header{}
+	header1 = types.Header{}
+	header2 = types.Header{}
+
+	block0 = types.Block{}
+
+	inputAddedEvent0 = inputbox.InputBoxInputAdded{}
+	inputAddedEvent1 = inputbox.InputBoxInputAdded{}
+	inputAddedEvent2 = inputbox.InputBoxInputAdded{}
+	inputAddedEvent3 = inputbox.InputBoxInputAdded{}
+
+	subscription0 = newMockSubscription()
+)
+
+type EvmReaderSuite struct {
+	suite.Suite
+	ctx        context.Context
+	cancel     context.CancelFunc
+	client     *MockEthClient
+	wsClient   *MockEthClient
+	inputBox   *MockInputBox
+	repository *MockRepository
+	evmReader  *EvmReader
+}
+
+func TestEvmReaderSuite(t *testing.T) {
+	suite.Run(t, new(EvmReaderSuite))
+}
+
+func (s *EvmReaderSuite) SetupSuite() {
+	s.ctx, s.cancel = context.WithTimeout(context.Background(), suiteTimeout)
+
+	err := json.Unmarshal(header0JsonBytes, &header0)
+	s.Require().Nil(err)
+	err = json.Unmarshal(header1JsonBytes, &header1)
+	s.Require().Nil(err)
+	err = json.Unmarshal(header2JsonBytes, &header2)
+	s.Require().Nil(err)
+
+	block0 = *types.NewBlockWithHeader(&header0)
+
+	err = json.Unmarshal(inputAddedEvent0JsonBytes, &inputAddedEvent0)
+	s.Require().Nil(err)
+	err = json.Unmarshal(inputAddedEvent1JsonBytes, &inputAddedEvent1)
+	s.Require().Nil(err)
+	err = json.Unmarshal(inputAddedEvent2JsonBytes, &inputAddedEvent2)
+	s.Require().Nil(err)
+	err = json.Unmarshal(inputAddedEvent3JsonBytes, &inputAddedEvent3)
+	s.Require().Nil(err)
+}
+
+func (s *EvmReaderSuite) TearDownSuite() {
+	s.cancel()
+}
+
+func (s *EvmReaderSuite) SetupTest() {
+
+	s.client = newMockEthClient()
+	s.wsClient = s.client
+	s.inputBox = newMockInputBox(s)
+	s.repository = newMockRepository()
+	inputReader := NewEvmReader(
+		s.client,
+		s.wsClient,
+		s.inputBox,
+		s.repository,
+		NodePersistentConfig{
+			DefaultBlock:            model.DefaultBlockStatusLatest,
+			InputBoxDeploymentBlock: 0,
+		},
+	)
+	s.evmReader = &inputReader
+}
+
+// Service tests
+func (s *EvmReaderSuite) TestItStopsWhenContextIsCanceled() {
+	ctx, cancel := context.WithCancel(s.ctx)
+	ready := make(chan struct{}, 1)
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- s.evmReader.Run(ctx, ready)
+	}()
+	cancel()
+
+	err := <-errChannel
+	s.Require().Equal(context.Canceled, err, "stopped for the wrong reason")
+}
+
+func (s *EvmReaderSuite) TestItEventuallyBecomesReady() {
+	ready := make(chan struct{}, 1)
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- s.evmReader.Run(s.ctx, ready)
+	}()
+
+	select {
+	case <-ready:
+	case err := <-errChannel:
+		s.FailNow("unexpected failure", err)
+	}
+}
```

Comment:
I did this but, on hindsight, I think we should generalize these tests and move them to https://github.com/cartesi/rollups-node/tree/next/2.0/internal/services